flake.nix

```
@@ -3,145 +3,61 @@
     liberaforms.flake = false;
-    mach-nix.url = "github:DavHau/mach-nix/3.5.0";
   };
 
-  outputs = {
-    self,
-    nixpkgs,
-    mach-nix,
-    ...
-  } @ inputs: let
-    # Extract version from VERSION.txt.
-    remove-newline = string: builtins.replaceStrings ["\n"] [""] string;
-    version = remove-newline (builtins.readFile (inputs.liberaforms + "/VERSION.txt"));
-
-    # Postgres setup script for tests.
-    initPostgres = ./nix/initPostgres.sh;
-
+  outputs = { self, nixpkgs, ... }@inputs:
+  let
     # System types to support.
     supportedSystems = ["x86_64-linux"];
     # Helper function to generate an attrset '{ x86_64-linux = f "x86_64-linux"; ... }'.
     genSystems = nixpkgs.lib.genAttrs supportedSystems;
     pkgsFor = nixpkgs.legacyPackages;
-  in {
-    overlays.default = _: prev: let
-      inherit (nixpkgs) lib;
 
-      req = builtins.readFile (inputs.liberaforms + "/requirements.txt");
-      # filter out "cryptography" as it makes mach-nix fail. also it is considered bad practice to hold back that package
-      filteredReq = lib.concatStringsSep "\n" (builtins.filter (e: e != "cryptography==36.0.1") (lib.splitString "\n" req));
+    resolveOverlays = first: overlays: with nixpkgs.lib; let
+      # Passing inputs here means overlays can be called with the standard signature
+      # but still be parametrized with flakes, maintaining composability. 
+      overlay' = flip (composeManyExtensions ([ (final: prev: { inherit inputs; }) ] ++ overlays));
+      # We don't to expose all of nixpkgs in the resulting attrset (like with .extend()), #TODO why not, or a .pkgs?
+      # so we take the fixpoint ourselves
+      attrs = fix (overlay' first);
+    in builtins.removeAttrs attrs [ "inputs" ]; #TODO I don't like doing it this way but I don't have a better idea how to keep the interface clean.
 
-      liberaforms-env = mach-nix.lib.${prev.system}.mkPython {
-        requirements = filteredReq;
-      };
-    in {
-      inherit liberaforms-env;
+    # TODO something is happening with argument processing, if I dont add the pkgs argument here it doesnt get passed to the module and breaks
+    # applyModuleArgsIfFunction /nix/store/43m6mis3zbnq5q9rw2yklnf6398p1x93-source/flake.nix
+    callModule = modPath: {pkgs, ...}@args: import modPath (args // { inherit inputs; });
 
-      liberaforms = prev.stdenv.mkDerivation rec {
-        inherit version;
-        pname = "liberaforms";
-
-        src = inputs.liberaforms;
-
-        dontConfigure = true; # do not use ./configure
-        propagatedBuildInputs = [liberaforms-env prev.postgresql] ++ (with prev.python39Packages; [
-          flask_migrate
-          flask_login
-          pillow
-        ]);
-
-        installPhase = ''
-          cp -r . $out
-        '';
-      };
-    };
+    mkSystem = system: modules: nixpkgs.lib.nixosSystem { inherit system modules; }; # TODO does the system argument make sense?
+  in {
+    overlays.default = import ./nix/overlay.nix;
```

Reference to comment about passing inputs to overlay, this is how you'd do it:
```suggestion
    overlays.default = import ./nix/overlay.nix inputs;
```